### PR TITLE
LocalModelCache Admission Webhook

### PR DESCRIPTION
**Pull request by HotsauceLee**
**What this PR does / why we need it**:
Adding this webhook to:
- Prevent deleting LocalModelCache when it's still being used
- Prevent creating new LocalModelCache CR with same storage URI

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

**1. DELETE test**

LocalModelCache CR:
```yaml
apiVersion: serving.kserve.io/v1alpha1
kind: LocalModelCache
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"serving.kserve.io/v1alpha1","kind":"LocalModelCache","metadata":{"annotations":{},"name":"iris"},"spec":{"modelSize":"1Gi","nodeGroup":"gpu1","sourceModelUri":"gs://kfserving-examples/models/sklearn/1.0/model"},"status":{"inferenceServices":[{"name":"sklearn-iris","namespace":"kserve"}]}}
  creationTimestamp: "2024-12-12T01:09:52Z"
  generation: 1
  name: iris
  resourceVersion: "319683"
  uid: 04318f7e-a8e4-4761-8f80-c25cf5a2745a
spec:
  modelSize: 1Gi
  nodeGroup: gpu1
  sourceModelUri: gs://kfserving-examples/models/sklearn/1.0/model
status:
  inferenceServices:
    - name: sklearn-iris
      namespace: kserve
```

isvc CR:
```yaml
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"serving.kserve.io/v1beta1","kind":"InferenceService","metadata":{"annotations":{},"name":"sklearn-iris","namespace":"kserve"},"spec":{"predictor":{"model":{"modelFormat":{"name":"sklearn"},"storageUri":"gs://kfserving-examples/models/sklearn/1.0/model"}}}}
  creationTimestamp: "2024-12-11T03:32:00Z"
  finalizers:
    - inferenceservice.finalizers
  generation: 1
  labels:
    internal.serving.kserve.io/localmodel: iris
  name: sklearn-iris
  namespace: kserve
  resourceVersion: "320037"
  uid: b9e9ad41-6b8d-48ed-95ff-c1898bfd1ba4
spec:
  predictor:
    model:
      modelFormat:
        name: sklearn
      name: ""
      resources: {}
      storageUri: gs://kfserving-examples/models/sklearn/1.0/model
```

When trying to delete the LocalModelCache CR:

```
yli1109@RVQWYXM093 kserve % k delete LocalModelCache iris                                                                                                                         
Error from server (Forbidden): admission webhook "localmodelcache.kserve-webhook-server.validator" denied the request: LocalModelCache iris is being used by InferenceService sklearn-iris
```

**2. CREATE test**

LocalModelCache CR:
```yaml
apiVersion: serving.kserve.io/v1alpha1
kind: LocalModelCache
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"serving.kserve.io/v1alpha1","kind":"LocalModelCache","metadata":{"annotations":{},"name":"iris"},"spec":{"modelSize":"1Gi","nodeGroup":"gpu1","sourceModelUri":"gs://kfserving-examples/models/sklearn/1.0/model"},"status":{"inferenceServices":[{"name":"sklearn-iris","namespace":"kserve"}]}}
  creationTimestamp: "2024-12-12T01:09:52Z"
  generation: 1
  name: iris
  resourceVersion: "319683"
  uid: 04318f7e-a8e4-4761-8f80-c25cf5a2745a
spec:
  modelSize: 1Gi
  nodeGroup: gpu1
  sourceModelUri: gs://kfserving-examples/models/sklearn/1.0/model
status:
  inferenceServices:
    - name: sklearn-iris
      namespace: kserve
```

New LocalModelCache CR:
```yaml
yli1109@RVQWYXM093 kserve % cat LocalModelCache_test2.yaml
apiVersion: serving.kserve.io/v1alpha1
kind: LocalModelCache
metadata:
  name: iris2
spec:
  modelSize: 1Gi
  nodeGroup: gpu1
  sourceModelUri: gs://kfserving-examples/models/sklearn/1.0/model
```

When trying to create the new LocalModelCache CR:
```
yli1109@RVQWYXM093 kserve % k apply -f LocalModelCache_test2.yaml
Error from server (Forbidden): error when creating "LocalModelCache_test2.yaml": admission webhook "localmodelcache.kserve-webhook-server.validator" denied the request: LocalModelCache iris has the same StorageURI gs://kfserving-examples/models/sklearn/1.0/model

```

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note

```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

### CODE REVIEW

File: pkg/webhook/admission/localmodelcache/local_model_cache_validation.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+// when it is created, updated, or deleted.
+//
+// NOTE: The +kubebuilder:object:generate=false and +k8s:deepcopy-gen=false marker prevents controller-gen from generating DeepCopy methods,
+// as this struct is used only for temporary operations and does not need to be deeply copied.
+type LocalModelCacheValidator struct {
+	client.Client
+}
+
+// +kubebuilder:webhook:verbs=delete,path=/validate-localmodelcaches,mutating=false,failurePolicy=fail,groups=serving.kserve.io,resources=localmodelcaches,versions=v1alpha1,name=localmodelcache.kserve-webhook-server.validator
+var _ webhook.CustomValidator = &LocalModelCacheValidator{}
```
Comment (yuzisun) [inline pkg/webhook/admission/localmodelcache/local_model_cache_validation.go:51]: The reason to use custom validator is to get the deletion event right?

File: pkg/webhook/admission/localmodelcache/local_model_cache_validation.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+func (v *LocalModelCacheValidator) ValidateDelete(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	localModelCache, err := convertToLocalModelCache(obj)
+	if err != nil {
+		localModelCacheValidatorLogger.Error(err, "Unable to convert object to LocalModelCache")
+		return nil, err
+	}
+	localModelCacheValidatorLogger.Info("validate delete", "name", localModelCache.Name)
+
+	// Check if current LocalModelCache is being used
+	for _, isvcMeta := range localModelCache.Status.InferenceServices {
```
Comment (yuzisun) [inline pkg/webhook/admission/localmodelcache/local_model_cache_validation.go:101]: Is it sufficient to check the size of the InferenceServices array assuming this status field is the "used isvcs"?

Comment (greenmoon55) [reply pkg/webhook/admission/localmodelcache/local_model_cache_validation.go:101]: @yuzisun This is more accurate. Is there a performance concern? e.g. if the model cache controller was not able to update the status field for some reason. like a bug.

File: pkg/webhook/admission/localmodelcache/local_model_cache_validation.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+		if err := v.Get(ctx, client.ObjectKey(isvcMeta), &isvc); err != nil {
+			localModelCacheValidatorLogger.Error(err, "Error getting InferenceService", "name", isvcMeta.Name)
+			return nil, err
+		}
+		modelName, ok := isvc.Labels[constants.LocalModelLabel]
+		if !ok {
+			continue
+		}
+		if modelName == localModelCache.Name {
+			return admission.Warnings{}, fmt.Errorf("LocalModelCache %s is being used by InferenceService %s", localModelCache.Name, isvcMeta.Name)
```
Comment (greenmoon55) [inline pkg/webhook/admission/localmodelcache/local_model_cache_validation.go:112]: noob question: what does the warning do? Would it reject the request?

File: pkg/webhook/admission/localmodelcache/local_model_cache_validation.go (hunk 2 above)
Comment (yuzisun) [reply pkg/webhook/admission/localmodelcache/local_model_cache_validation.go:101]: are you saying for some reason the used inference services are not removed from status?

Comment (greenmoon55) [reply pkg/webhook/admission/localmodelcache/local_model_cache_validation.go:101]: for example, the local model controller keep crashing, then maybe the service was not added to or removed from the status field. But yeah, not a big difference.

File: pkg/webhook/admission/localmodelcache/local_model_cache_validation.go (hunk 1 above)
Comment (HotsauceLee) [reply pkg/webhook/admission/localmodelcache/local_model_cache_validation.go:51]: are you referring to `kubebuilder:webhook:verbs=delete` ? I think it determines which event to capture. Custom validators need to implement 3 methods(create, update and delete) and I only saw logs for delete

File: pkg/webhook/admission/localmodelcache/local_model_cache_validation.go (hunk 2 above)
Comment (HotsauceLee) [reply pkg/webhook/admission/localmodelcache/local_model_cache_validation.go:101]: is the concern there might be isvc using this not in the array? imo inactive isvc not removed from the array is more likely

File: pkg/webhook/admission/localmodelcache/local_model_cache_validation.go (hunk 3 above)
Comment (HotsauceLee) [reply pkg/webhook/admission/localmodelcache/local_model_cache_validation.go:112]: yes it rejects the request. Just updated the PR description with my test, you can see rejection there. Also refer to [this](https://github.com/kserve/kserve/blob/master/pkg/apis/serving/v1beta1/inference_service_validation.go#L106-L127)

**Review by yuzisun — APPROVED** [review]
/lgtm
/approve

**Pull request merged**